src/lib/string_pool.rs

```diff
@@ -0,0 +1,509 @@
+        &self,
+        enc: &ENCODING,
+        mut read_buf: ExpatBufRef,
+        vec: &mut BumpVec<XML_Char>,
+    ) -> bool {
+        let start = vec.as_ptr();
+
+        // REVIEW: Can this be replaced with self.is_empty() &&?
+        if start.is_null() && !self.rented_grow(vec) {
+            return false;
+        }
+
+        let mut cap = vec.capacity();
+        let mut len = vec.len();
+
+        loop {
+            debug_assert!(len <= cap);
+
+            // Continue to allocate if we don't have enough space
+            while (cap - len) < read_buf.len() {
+                if !self.rented_grow(vec) {
+                    return false;
+                }
+
+                cap = vec.capacity();
+            }
+
+            let start_len = vec.len();
+            let cap = vec.capacity();
+
+            vec.resize(cap, 0);
+
+            let mut write_buf = unsafe { ExpatBufRefMut::new_len(vec.as_mut_ptr().add(len), cap - start_len) };
+            let write_buf_len = write_buf.len();
+            let convert_res = unsafe {
+                XmlConvert!(enc, &mut read_buf, &mut write_buf)
+            };
+            // The write buf shrinks by how much was written to it
+            let written_size = write_buf_len - write_buf.len();
+
+            vec.truncate(start_len + written_size);
+
+            if convert_res == XML_Convert_Result::COMPLETED || convert_res == XML_Convert_Result::INPUT_INCOMPLETE {
+                break;
+            }
+        }
+
+        true
+    }
+
+    pub(crate) unsafe fn copy_c_string(
+        &self,
+        mut s: *const XML_Char,
+    ) -> Option<&mut [XML_Char]> {
+        // self.append_c_string(s);?
+        let successful = self.inner().rent(|vec| {
+            let mut vec = vec.borrow_mut();
+
+            loop {
+                if !self.rented_append_char(*s, &mut *vec) {
+                    return false;
+                }
+                if *s == 0 {
+                    break;
+                }
+                s = s.offset(1);
+            }
+
+            true
+        });
+
+        if !successful {
+            return None;
+        }
+
+        Some(self.finish_string())
+    }
+
+    pub(crate) unsafe fn copy_c_string_n(
+        &self,
+        mut s: *const XML_Char,
+        mut n: c_int,
+    ) -> Option<&[XML_Char]> {
+        let successful = self.inner().rent(|vec| {
+            let mut vec = vec.borrow_mut();
+
+            while n > 0 {
+                if !self.rented_append_char(*s, &mut vec) {
+                    return false;
+                }
+                n -= 1;
+                s = s.offset(1)
+            }
+
+            true
+        });
+
+        if !successful {
+            return None;
+        }
+
+        Some(&*self.finish_string())
+    }
+
+    pub(crate) fn grow(&self) -> bool {
+        self.inner().rent(|vec| self.rented_grow(&mut *vec.borrow_mut()))
+    }
+
+    fn rented_grow(&self, vec: &mut BumpVec<XML_Char>) -> bool {
```

Comment:
Doesn't `BumpVec` have its own exponential doubling allocation strategy? We could get rid of ours and always use theirs. Calling `try_reserve` uses it internally.

Reply:
In other words, I feel like we shouldn't need this whole "check length & grow" pattern if we're using `BumpVec` which already does all of that internally. `try_reserve`, `push` and `extend` should be enough for our needs.

Reply:
Maybe you're right, but it does feel like it's something we can clean up later?